### PR TITLE
Adds support for saving and restoring pane titles.

### LINKED ISSUE #330: Pane titles are not restored
Names that are set using `select-pane -T [name]` are not restored at all.

### PR DESCRIPTION
**Pull request by Hologos**
This PR is a rebase of #342 with addressed PR comments, fixes for `restore_shell_history()` and `dump_shell_history()`. It fixes #330 as well.

I am currently testing it but pls test it thoroughly yourself.

It is a breaking change so I added a comment to the changelog.

### CODE REVIEW

**Comment by Hologos** [comment]
TODOs:

- [ ] add pane titles into test fixture files
- [ ] BUG: fix setting pane title when pane already exists

**Comment by Hologos** [comment]
> Instead of naming the pane with a standalone command, can we somehow set the correct pane_title when the pane is created?
>
> _Originally posted by @bruno- in https://github.com/tmux-plugins/tmux-resurrect/pull/342#discussion_r815397808_

Sadly it's not possible, I created a function that is called within `new_pane()` and `restore_pane()` (if pane already exists) in
 4a3b59c84bff946348f67e5919dd32e7e8bf3f61. Let me know if I should revert to calling it like in the original PR.

**Comment by Hologos** [comment]
Last commit reverts to setting `pane_title` at the end of `restore_pane()`. Otherwise it has to be called in `new_pane()`, `new_window()` and `new_session()` and once again after the very first pane is killed. This covers all cases.

File: scripts/restore.sh [hunk 1]
```diff
@@ -302,7 +312,7 @@ restore_window_properties() {
 }
 
 restore_shell_history() {
-	awk 'BEGIN { FS="\t"; OFS="\t" } /^pane/ { print $2, $3, $6, $9; }' $(last_resurrect_file) |
+	awk 'BEGIN { FS="\t"; OFS="\t" } /^pane/ { print $2, $3, $6, $10; }' $(last_resurrect_file) |
```
Comment (bruno-) [inline scripts/restore.sh:315]: I'm trying to get rid of this feature for a long time

https://github.com/tmux-plugins/tmux-resurrect/blob/master/docs/restoring_shell_history.md

Please don't fix this, I'll remove "restore shell history" after we merge restoring pane titles feature.

File: scripts/restore.sh [hunk 2]
```diff
@@ -200,6 +209,7 @@ restore_pane() {
 		else
 			new_session "$session_name" "$window_number" "$dir" "$pane_index"
 		fi
+		set_pane_title "$session_name" "$window_number" "$pane_index" "$pane_title"
```
Comment (bruno-) [inline scripts/restore.sh:212]: `set_pane_title` function effectively contains only a single line/single command, and it's called from just one place. 

My first thought was why not just inline it? If you think the function name makes it obvious what the cryptic tmux command does, we can also add a comment.

But, if you want to keep the function, that's fine too.

**Comment by bruno-** [comment]
@Hologos thank you for working on this.

> Let me know if I should revert to calling it like in the original PR.

I'd prefer if we inlined that function, but I'm ok if you really want to keep `set_pane_title`.

I left a couple small comments, but overall this looks good to me.
Just to confirm, you ran this branch on your machine and everything looks good?

Can you please ping me after you make the final tweaks? I'll then test this myself for a couple days, and then we can merge.

File: scripts/restore.sh (hunk 1 above)
Comment (Hologos) [reply scripts/restore.sh:315]: Does it mean I should revert all changes in `restore_shell_history()` as well as `dump_shell_history()`?

File: scripts/restore.sh (hunk 2 above)
Comment (Hologos) [reply scripts/restore.sh:212]: It is a remnant of previous commits where I tried to call it when creating panes. I agree to just inline the command.

**Comment by Hologos** [comment]
> Just to confirm, you ran this branch on your machine and everything looks good?
> 
> Can you please ping me after you make the final tweaks? I'll then test this myself for a couple days, and then we can merge.

I've been using it for a week now, resurrected several times + tested it the day I made the changes. We can test it for longer, I am able to rebase it if any changes gets merged into master.

File: scripts/restore.sh (hunk 1 above)
Comment (bruno-) [reply scripts/restore.sh:315]: Yes please.

I'll remove the whole "shell history" feature after we merge this branch.

**Comment by bruno-** [comment]
> I've been using it for a week now, resurrected several times + tested it the day I made the changes.

Great. I'll start using it after you make the final tweaks. We can merge this after a couple days if all looks good.

**Comment by Hologos** [comment]
I've made requested changes. Thank you for your time.

**Comment by bruno-** [comment]
Thanks.

I pulled this branch locally. Testing basic save/restore looks good. I'll continue working with it for the next couple days.

**Comment by bruno-** [comment]
@Hologos this PR is ready for the merge.

Since this feature contains breaking changes I'll have to release a new major version. That's done by tagging the commit with breaking changes with a new major version `v4.0.0`. To do that, the feature would ideally be one commit.

So, can you please squash the commits from this branch in a single commit? Thanks

**Comment by Hologos** [comment]
@bruno- Hi, I squashed them into a single one. Cheers

**Comment by ioogithub** [comment]
Thank you for working on this feature. Is this ready to test yet?

**Comment by ioogithub** [comment]
I would love to try this. I see version 4.0 was release 16 days ago. Does this version contain the change? I did not see a changelog on the release page.

**Comment by Hologos** [comment]
@ioogithub The feature is part of v4.0 release.

**Pull request merged**